src/chunk.c

```diff
@@ -1123,6 +1153,62 @@ chunk_scan_ctx_get_chunk(ChunkScanCtx *ctx)
 	return ctx->data;
 }
 
+/*
+ * Resurrect a chunk from a tombstone.
+ *
+ * A chunk can be dropped while retaining its metadata as a tombstone. Such a
+ * chunk is marked with dropped=true.
+ *
+ * This function resurrects such a dropped chunk based on the original metadata,
+ * including recreating the table and related objects.
+ */
+static Chunk *
+chunk_resurrect(Hypertable *ht, ChunkStub *stub)
+{
+	ScanIterator iterator;
+	Chunk *chunk = NULL;
+	int count = 0;
+
+	Assert(NULL != stub->constraints);
+	Assert(NULL != stub->cube);
+
+	iterator = ts_scan_iterator_create(CHUNK, RowExclusiveLock, CurrentMemoryContext);
+	init_scan_by_chunk_id(&iterator, stub->id);
+
+	ts_scanner_foreach(&iterator)
+	{
+		TupleInfo *ti = ts_scan_iterator_tuple_info(&iterator);
+		HeapTuple new_tuple;
+
+		Assert(count == 0 && chunk == NULL);
+		chunk = palloc0(sizeof(Chunk));
+		chunk_formdata_fill(&chunk->fd, ti->tuple, ti->desc);
+		chunk->fd.dropped = false;
+		new_tuple = chunk_formdata_make_tuple(&chunk->fd, ti->desc);
+		ts_catalog_update_tid(ti->scanrel, &ti->tuple->t_self, new_tuple);
+		heap_freetuple(new_tuple);
+
+		/* Fill in the stub information and create data table and related
+		 * objects while we hold lock on the metadata row */
+		chunk->constraints = stub->constraints;
+		chunk->cube = stub->cube;
+		chunk_create_table_after_lock(ht, chunk);
+		count++;
+
+		/* Assume there's only one match. (We break early to avoid scanning
+		 * also the updated tuple.) */
+		break;
+	}
+
+	ts_scan_iterator_close(&iterator);
+
+	Assert(count == 0 || count == 1);
+
```

Comment:
this functions is called when we see the state as dropped. Something is wrong if we try to reconstruct the metadata and it has been deleted from under us. This should error out and not silently ignore.
I am not clear about the changes to this function. Should discuss this further.

Reply:
Can you explain why it is wrong?

This is my reasoning: with read committed isolation, the current transaction would be able to see changes that another transaction has made as soon as the other transaction is committed. If that other transaction deleted the chunk metadata (e.g., because the continuous aggregate was dropped), then the chunk metadata would be gone by the time we scan it in this function unless we held a lock on the chunk row from the point we created the ChunkStub that is an input parameter to this function.

In other words, how can we guarantee that the the metadata we used to create the ChunkStub remains until we scan it again in this function?

We actually have similar logic when creating the chunk initially to protect against a case when another transaction races to create the chunk before us. We first do a cheap lookup for an existing chunk without grabbing a heavy lock, then if the chunk is not found, we grab the heavy lock (on hypertable) followed by another lookup to ensure that the chunk hasn't been created by someone else before we grabbed the lock.

Thus, the logic is similar here: when we have determined we need to resurrect a chunk, we need to grab a heavy lock (RowExclusiveLock) on the row, but if the row is not found it must mean someone else deleted it before we got the lock. Or, am I missing something?

Note that we still hold the hypertable lock here, but I don't think it protects us here since the path to delete is different from the path to create.

Reply:
@gayyappan is this resolved now?

Reply:
yes.